### PR TITLE
executor/show: show create table return utf8mb4 charset.

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
Old TiDB `show create table` will return charset by user specified, but actually treat all the data as utf8mb4. 

This will cause some problems when drainer to mysql.
eg:
```sql
#In tidb
mysql> create table t2 (a char(50)) charset=utf8;
Query OK, 0 rows affected
mysql> insert into t2 values (UNHEX("F09F8C9F"));
Query OK, 1 row affected

#In MySQL:
mysql> create table t2 (a char(50)) charset=utf8;
Query OK, 0 rows affected
Time: 0.011s
mysql> insert into t2 values (UNHEX("F09F8C9F"));
(1366, u"Incorrect string value: '\\xF0\\x9F\\x8C\\x9F' for column 'a' at row 1")
```

### What is changed and how it works?
This PR make `show create table` charset always return `utf8mb4`, no matter what charset the user specified.
And warning when create table if table charset is not `utf8mb4`.
```sql
mysql root@127.0.0.1:test> create table t2 (a char(50)) charset=utf8;
Query OK, 0 rows affected
Time: 0.023s
mysql root@127.0.0.1:test> show warnings;
+---------+------+---------------------------------------------------------------------------------+
| Level   | Code | Message                                                                         |
+---------+------+---------------------------------------------------------------------------------+
| Warning | 1105 | TiDB only supports the "utf8mb4" character set, so "utf8" does not take effect. |
+---------+------+---------------------------------------------------------------------------------+

```

### Check List 

Tests 

Code changes

Side effects

Related changes

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8604)

### CODE REVIEW

File: executor/show.go [hunk 1]
```diff
@@ -662,24 +662,10 @@ func (e *ShowExec) fetchShowCreateTable() error {
...
-	if len(collate) == 0 {
-		// If we can not find default collate for the given charset,
-		// do not show the collate part.
-		fmt.Fprintf(&buf, " DEFAULT CHARSET=%s", charsetName)
-	} else {
-		fmt.Fprintf(&buf, " DEFAULT CHARSET=%s COLLATE=%s", charsetName, collate)
-	}
+	// Currently TiDB treat all the data as utf8mb4 actually.
+	// Make the TiDB's query result consistent with its actual behavior.
+	charsetName, collate := charset.GetDefaultCharsetAndCollate()
```
[the remark this answers is not in the document]
Comment (jackysp) [reply executor/show.go]: Both the `create table` and `show create table` need a warning?

**Comment by jackysp** [comment]
Please fix CI @crazycs520 .

File: executor/show.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (crazycs520) [reply executor/show.go]: Done. em..., when user specified charset is utf8, should append warning?
BTW, Should we also append warning when create table or alter table charset when user specified charset is not utf8mb4?

[the remark this answers is not in the document]
Comment (winkyao) [reply executor/show.go]: when charset is utf8, should return warning.

[the remark this answers is not in the document]
Comment (winkyao) [reply executor/show.go]: @crazycs520 I means, when the client create table with not an utf8/utf8mb4 charset, we should return warning. not here.

[the remark this answers is not in the document]
Comment (crazycs520) [reply executor/show.go]: done.

**Comment by gregwebs** [comment]
I haven't had a chance to look at the patch in detail, but I like the concept of only accepting utf8mb4 and providing a warning if utf8 is given. For other charsets outside of utf8/utf8mb4 it would be best to be an error. I know we are worried about someone's program breaking, but I think it might be rare and it is better to break upfront rather than later on. We would probably still have to just provide a warning for latin1 since it is the MySQL 5 default.

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -958,6 +958,10 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
 	if err != nil {
 		return errors.Trace(err)
 	}
+	if len(tbInfo.Charset) != 0 && strings.ToLower(tbInfo.Charset) != mysql.DefaultCharset {
+		ctx.GetSessionVars().StmtCtx.AppendWarning(errors.Errorf("The user specified charset of table '%s.%s' is %s, currently TiDB treat all the data as utf8mb4 actually.", schema.Name, tbInfo.Name, tbInfo.Charset))
```
Comment (jackysp) [inline ddl/ddl_api.go]: It's better to refine this sentence because it is important for the user. PTAL @lilin90 .

**Comment by zz-jason** [comment]
@crazycs520 I think the PR description should be updated.

File: ddl/ddl_api.go (hunk 2 above)
Comment (lilin90) [inline ddl/ddl_api.go]:
```suggestion
		ctx.GetSessionVars().StmtCtx.AppendWarning(errors.Errorf("The user specified charset of table '%s.%s' is %s. Currently, TiDB treats all the data as utf8mb4 actually.", schema.Name, tbInfo.Name, tbInfo.Charset))
```

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -958,6 +958,10 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, s *ast.CreateTableStmt) (err e
 	if err != nil {
 		return errors.Trace(err)
 	}
+	if len(tbInfo.Charset) != 0 && strings.ToLower(tbInfo.Charset) != mysql.DefaultCharset {
+		ctx.GetSessionVars().StmtCtx.AppendWarning(errors.Errorf("The user specified charset of table '%s.%s' is %s. Currently TiDB treats all the data as utf8mb4 actually.", schema.Name, tbInfo.Name, tbInfo.Charset))
```
Comment (jackysp) [inline ddl/ddl_api.go]: How about
"TiDB only supports setting the character set as utf8mb4, %s does not take effect." ?

Comment (lilin90) [reply ddl/ddl_api.go]: Inappropriate in grammar. Change to:

"TiDB only supports setting the character set as utf8mb4, **so** %s does not take effect."

**Comment by lysu** [comment]
@crazycs520 this pr doesn't pass all test case

**Pull request merged**